x-pack/plugins/case/server/services/tags/read_tags.ts

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { SavedObject, SavedObjectsClientContract } from 'kibana/server';
+import { CASE_SAVED_OBJECT } from '../../constants';
+import { CaseAttributes } from '../..';
+
+const DEFAULT_PER_PAGE: number = 1000;
+
+export const convertToTags = (tagObjects: Array<SavedObject<CaseAttributes>>): string[] =>
+  tagObjects.reduce<string[]>((accum, tagObj) => {
+    if (tagObj && tagObj.attributes && tagObj.attributes.tags) {
+      return [...accum, ...tagObj.attributes.tags];
+    } else {
+      return accum;
+    }
+  }, []);
+
+export const convertTagsToSet = (tagObjects: Array<SavedObject<CaseAttributes>>): Set<string> => {
+  return new Set(convertToTags(tagObjects));
+};
+
+// Note: This is doing an in-memory aggregation of the tags by calling each of the alerting
+// records in batches of this const setting and uses the fields to try to get the least
+// amount of data per record back. If saved objects at some point supports aggregations
+// then this should be replaced with a an aggregation call.
+// Ref: https://www.elastic.co/guide/en/kibana/master/saved-objects-api.html
+export const readTags = async ({
+  client,
+  perPage = DEFAULT_PER_PAGE,
+}: {
+  client: SavedObjectsClientContract;
+  perPage?: number;
+}): Promise<string[]> => {
+  const tags = await readRawTags({ client, perPage });
+  return tags;
+};
+
+export const readRawTags = async ({
+  client,
+  perPage = DEFAULT_PER_PAGE,
+}: {
+  client: SavedObjectsClientContract;
+  perPage?: number;
+}): Promise<string[]> => {
+  const firstTags = await client.find({
+    type: CASE_SAVED_OBJECT,
+    fields: ['tags'],
+    page: 1,
+    perPage,
+  });
+  const firstSet = convertTagsToSet(firstTags.saved_objects);
+  const totalPages = Math.ceil(firstTags.total / firstTags.per_page);
+  if (totalPages <= 1) {
+    return Array.from(firstSet);
+  } else {
+    const returnTags = await Array(totalPages - 1)
+      .fill({})
+      .map((_, page) => {
+        // page index starts at 2 as we already got the first page and we have more pages to go
+        return client.find({ type: CASE_SAVED_OBJECT, fields: ['tags'], perPage, page: page + 2 });
```

Comment:
oh, we had a bad bug here in detection with this logic. let's talk about it.

Reply:
is it updated/fixed in detection? stole this several weeks ago